[not for merge] Trinitus01 changes to amazonechocontrol

J-N-K opened the pull request:
<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/development/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/development/bindings.html#static-code-analysis
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

Trinitus01 committed:
added: channel for account announcements (to all devices)
added: channel to play sound of Alexa Skills Kit Sound Library
added: stacked or grouped announcements (too many request fix)
added: sequenced stacked or grouped textToSpeech (too many request fix)
added: ttsvolume in stacked or grouped announcements & sequenced textToSpeech (too many request fix)
removed: stacked or grouped textToSpeech converted to announcements
fixed: update or refresh volume immediately after a command (before updated on second command and broke ttsvolume, command, volume sequence)
fixed: update or refresh volume immediately after manual echo button press (before did not update or refresh anyway)
fixed: do not show xml tags on echos with display during announcements

Trinitus01 committed:
added: description of new features to the README
changed: the description for the thing type sound

J-N-K commented:
@Trinitus01 Doesn't look that bad. Some general comments: 

- The `.properties` file has wrong encoding, ISO-8859-1 is needed here, it looks like you used UTF-8
- `@author`-tag needs a real name, for a new file like `@author John Doe - Initial Contribution
- Some formatting looks wrong. Please apply spotless-formatter (`mvn spotless:apply -pl bundles\org.openhab.binding.amazonechocontrol`) 

I guess most conflicts should be solved then.

Trinitus01 commented:
Thanks for the feedback i will take a look into it.
I also forgot to add the "Signed-off-by:" to the commit (else i read it is necessary), will this be a problem?
So maybe i have to do a commit amend and a push force or do the fork stuff again :(

I realised that the new sound channel is not necessary because someone on the community wrote he implemented it via SSML but in the new version it doesn't work anymore. I also tried to do some days ago and thought it is not possible there is a difference between speak and sound. Now i know it is/was possible and i would check this too.

J-N-K commented:
We can fix the sign-off later, no need to worry.

Trinitus01 commented:
Today i found a solution for parallel tts (the known way is sequenced) so now it is possible to send one tts to all echos with only one request (i thought this is only possible via announcement, like michi also thought i think). A sequenced tts arrives the echos step by step and produce a resound effect, very bad, but a parallel tts is like you send x requests to every single echo at the same time, but in one request. This also enables ssml audio aso. The best thing is, for the user there is no change in usage anymore. (Because of the version major changes aso) The user can use announcement like before and tts like before without the too many request issue. There was also an issue with the web socket because if you send a parallel tts, a lot of web socket messages produces a lot of requests again. I fixed it, so at the moment it works very well. Currently i am testing all the stuff. After testing, i will fix the issues you reported and do a push. Hopefully tomorrow or maybe the day after.

J-N-K commented:
Unfortunately I had to do some refactoring to remove potential bugs. We‘ll see if we can simply merge or need some more detailed work.

Trinitus01 commented:
Maybe you can help me, i have the following problem: a method is called many times in one second and i want to queue them and call them with a delay of x seconds. The first call should run the code immediately and the second, third, aso should called delayed. If the delay runs out, it should call immediately again. Any idea how to implement this. In real life it means for example: 3 commands with the same text, lets say "hello" are send to different echo items and 1 command with another text lets say "hello too" is send to the different echo items too (1 second next to the first one). For now only "hello too" is spoken because the second request overlap the first one. A delay is useless because each sentence has another length. So the only way is to send the tts, receive the tts is arrived via websocket and unlock the tts for the next request: method1 send tts, lock and wait for method2, method2 wait for receiving tts via websocket and unlock method1, now method2 can call the second tts, and so on. I will find a solution for sure but maybe you have an idea and save some time^^

Trinitus01 commented:
Wrong desciption correction:  method1 send tts, lock and wait for method2, method2 wait for receiving tts via websocket and unlock method1, now !method1! can call the second tts, and so on. I will find a solution for sure but maybe you have an idea and save some time^^

J-N-K commented:
Something like this:

```

    private LinkedBlockingQueue<String> messageQueue = new LinkedBlockingQueue<>();
    private AtomicBoolean messageQueueRunning = new AtomicBoolean();
    private ScheduledFuture<?> messageSenderUnblockFuture;

    // call this to send a message
    public void queueMessage(String message) {
        messageQueue.add(message);
        if (messageQueueRunning.compareAndSet(false,true)) {
            processQueue();
        }
    }

    // called by the websocket when the response is received or after 30s by the scheduled unblocker
    public void messageSent() {
        ScheduledFuture<?> messageSenderUnblockFuture = this.messageSenderUnblockFuture;
        if (messageSenderUnblockFuture != null) {
            messageSenderUnblockFuture.cancel(true);
            this.messageSenderUnblockFuture = null;
        }
        if (!messageQueue.isEmpty()) {
            processQueue();
        } else {
            messageQueueRunning.set(false);
        }
    }

    private void processQueue() {
        String nextMessage = messageQueue.poll();

        // do whatever is needed to send the message

        // set a timeout, if we do not receive the websocket response within 30s, unblock the queue
        messageSenderUnblockFuture = scheduler.schedule(this::messageSent, 30, TimeUnit.SECONDS);
    }
```

Trinitus01 commented:
Yes, your example looks very near to what i am looking for! I will try to implement it and give a feedback. Very big thanks!

Trinitus01 commented:
Your example work like it should and its a very nice to handle a sync process, i will take this as a big lesson. But, sadly, amazon send the web socket as soon as the request arrives. So method1 call a tts and method2 immediately receives that the tts is called and calls method1 again. (like it should) But the time between the method1 calls are too short and this produces a "too many"-ex. It seems there is no other solution than set a time waiting for the next request. But the time has to be longer then the speaking time. Else it ends in something like: Sentence1 "This is a test and the test takes a lot of time", Sentence2 "And this is another test", Result "This ist a test and the...is another test" (Overlap). If it is not possible to recognise if a tts is finished, there is no possibility to start the next one after the first one :( In my case i use a tts to say: "xyz is in front of the house door" and if the door opens "the front door is going to open". But they overlap and i only receive one or a half of them. One idea is to count the words and set 1 second for every word, but this is -for sure- bumbling.

J-N-K commented:
```
 private LinkedBlockingQueue<String> messageQueue = new LinkedBlockingQueue<>();
    private AtomicBoolean messageQueueRunning = new AtomicBoolean();
    private ScheduledFuture<?> messageSenderUnblockFuture;

    // call this to send a message
    public void queueMessage(String message) {
        messageQueue.add(message);
        if (messageQueueRunning.compareAndSet(false,true)) {
            processQueue();
        }
    }

    private void processQueue() {
        if (!messageQueue.isEmpty()) {
            String nextMessage = messageQueue.poll();

            // do whatever is needed to send the message

            int wordcount = nextMessage.split("\\s+").length;

            // process next
            messageSenderUnblockFuture = scheduler.schedule(this::messageSent, wordcount, TimeUnit.SECONDS);
        } else {
            messageQueueRunning.set(false);
            messageSenderUnblockFuture = null;
        }
    }

```

Maybe like this. But don't forget to check the future in `dispose` and cancel it if non-null, otherwise we are leaking resources again.

Trinitus01 commented:
Hi,

with your code i was able to fix it. But i used the count of all letters in a sentence instead the words. 100 milliseconds per letter seems to be a perfect value. No overlapping anymore, nearly same delays between sentences, very nice. Should i test it for some days or should i push it so you are able to take a look at? I still have to fix the encoding and format.

Br

Trinitus01 committed:
changed: .properties encoding to ISO-8859-1
changed: @author-tag to real name
applied: apply spotless-formatter

changed: sequenced tts to parallel tts
fixed: parallel websocket messages produces too many requests (queued delay websocket messages)
fixed: overlapping tts (queued delay tts messages)
removed: sound channel, better to use tts ssml
fixed: do not convert a tts to an announcement if ssml is used, tts can also handle ssml

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 commented:
Pushed.
It seems there are only some little conflicts left.

Br

Trinitus01 commented:
Btw there is a description in the README how to use a json for tts (called Expert) with an option sound:

{ "sound": true, "speak":"<Speak>" "title": "<Title>", "body": "<Body Text>", "volume": 20}

This "sound"-option is useless, it is only used to check if sound true and ssml speak is not allowed but thats all. the sound parameter is never used in any functional way.

Br

Trinitus01 committed:
changed: try to reduce some conflicts (the easy stuff)

Trinitus01 committed:
changed: try to reduce some conflicts (the easy stuff) (2)

Trinitus01 committed:
changed: try to reduce some conflicts (the easy stuff) (3)

J-N-K commented on bundles/org.openhab.binding.amazonechocontrol/README.md:
````diff
@@ -180,6 +182,10 @@ Sample for the Thing echo1 only. But it will work in the same way for the other
 ```
+// Account
+String Echo_Living_Room_SendMessage            "SendMessage"                           {channel="amazonechocontrol:account:account1:sendMessage"}
+String Echo_Living_Room_AccountAnnouncement    "AccountAnnouncement"                   {channel="amazonechocontrol:account:account1:accountAnnouncement"}
````
Shouldn't the cannel be named `sendAnnouncement` to keep it consistent with the `sendMessage`?

J-N-K reviewed (commented):
There are some more comments. I wouldlike to address them by creating a PR to your repo. Can you enable that?

Am I right that this is now fully backward compatible?

Trinitus01 commented:
Yes it should be fully backward compatible. But keep in mind that i dont have that much use cases like other people. On the base of code i see no difference in usage.

I will change accountAnnouncement to sendAnnouncement tomorrow in the morning.

How to enable "it"? Maybe u can tell me else i will google it tomorrow in the morning, too.

Br

Trinitus01 commented:
Found an issue :( It seems textToSpeechQueueRunning (the AtomicBoolean) stay true and queuedTextToSpeech is never called again (tts stop working). So i still have to take a closer look at it.

Trinitus01 commented:
I added a "Branch Protection Rule" to * Branches with the options "Require pull request reviews before merging" and "Require review from Code Owners". Was it the correct thing to enable?

Br

Trinitus01 committed:
changed: channel name accountAnnouncement to sendAnnouncement

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 commented:
Hi,

sorry to ask you for help again :(

I queued tts and web socket but i realised the solution is not to queue the individual functions.
The real major problem is the function "makeRequest" in the "Connection" class.

Every feature like tts, announcement, web socket, aso use this single function to send a request, for sure.
The knowledge is to queue and delay !this! single function and not the functions for tts, ann...aso.
To solve the problem for tts and websocket i used your example, but its not possible for the "makeRequest" function, because it returns a value and is not at void function.
Now, i really don't know how to queue and delay the "makeRequest" function, i tried and i failed.

btw this is the big hole of "too many requests"-issue, every thing else was only a "hole"-fix for the single functions. My current version is stable but its not perfect because the major fix is missing. The structure of the binding in this case is not well, hopefully you can give me a hint to solve this problem.

Br

Trinitus01 committed:
fixed: unhandled exception stop tts
fixed: try to reduce crossing requests because of too many requests

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 committed:
added: ttsVolume for every single device (tts)
fixed: lastKnownVolume and ttsVolume issue (tts)

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 committed:
added: ttsVolume for every single device (announcement)
fixed: lastKnownVolume and ttsVolume issue (announcement)

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 committed:
fixed: ttsVolume for announcements

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 committed:
removed: channel for account announcements (better to use grouped or stacked announcements)

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 committed:
fixed: check for empty text (tts and announcement)
changed: the way to extract text from xml (before it did not work for less than 3 chars)

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>

Trinitus01 committed:
removed: debug information logging
fixed: finally fixed crossing requests
fixed: ttsVolume and lastKnownVolume (tts and announcement) (was still broken before)

Signed-off-by: Tom Blum (Trinitus01) <trinitus01@googlemail.com>
